### PR TITLE
Hide niches under UnsafeCell

### LINKED ISSUE #68206: Mutex and RwLock are unsound in presence of discriminant elision
Neither Mutex type nor RwLock type inhibit layout optimization and so either of
them might contain a niche. When a niche is used to perform a discriminant
elision extracting the discriminant of the outer type will inadvertently load
the content of Mutex / RwLock without proper synchronization.

### LINKED ISSUE #68303: miri error in safe code when mutating the bool inside Option<RefCell<bool>>
Running this code in miri(2019-12-27) in the playground:
```rust
use std::cell::RefCell;

fn main() {
    let optional=Some(RefCell::new(false));
    let mut handle=optional.as_ref().unwrap().borrow_mut();
    optional.is_some();
    *handle=true;
}
```
https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=3e8c5326890889354418d3161294d0fe

Results in this MIRI error:
```

error: Miri evaluation error: trying to reborrow for Unique, but parent tag <1307> does not have an appropriate item in the borrow stack
    --> /root/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libcore/cell.rs:1388:9
     |
1388 |         self.value
     |         ^^^^^^^^^^ Miri evaluation error: trying to reborrow for Unique, but parent tag <1307> does not have an appropriate item in the borrow stack
     |
note: inside call to `<std::cell::RefMut<bool> as std::ops::DerefMut>::deref_mut` at src/main.rs:7:5
    --> src/main.rs:7:5
     |
7    |     *handle=true;
     |     ^^^^^^^
     = note: inside call to `main` at /root/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libstd/rt.rs:67:34
     = note: inside call to closure at /root/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libstd/rt.rs:52:73
     = note: inside call to closure at /root/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libstd/sys_common/backtrace.rs:129:5
     = note: inside call to `std::sys_common::backtrace::__rust_begin_short_backtrace::<[closure@DefId(1:6016 ~ std[49a3]::rt[0]::lang_start_internal[0]::{{closure}}[0]::{{closure}}[0]) 0:&dyn std::ops::Fn() -> i32 + std::marker::Sync + std::panic::RefUnwindSafe], i32>` at /root/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libstd/rt.rs:52:13
     = note: inside call to closure at /root/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libstd/panicking.rs:296:40
     = note: inside call to `std::panicking::r#try::do_call::<[closure@DefId(1:6015 ~ std[49a3]::rt[0]::lang_start_internal[0]::{{closure}}[0]) 0:&&dyn std::ops::Fn() -> i32 + std::marker::Sync + std::panic::RefUnwindSafe], i32>` at /root/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libstd/panicking.rs:272:13
     = note: inside call to `std::panicking::r#try::<i32, [closure@DefId(1:6015 ~ std[49a3]::rt[0]::lang_start_internal[0]::{{closure}}[0]) 0:&&dyn std::ops::Fn() -> i32 + std::marker::Sync + std::panic::RefUnwindSafe]>` at /root/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libstd/panic.rs:394:14
     = note: inside call to `std::panic::catch_unwind::<[closure@DefId(1:6015 ~ std[49a3]::rt[0]::lang_start_internal[0]::{{closure}}[0]) 0:&&dyn std::ops::Fn() -> i32 + std::marker::Sync + std::panic::RefUnwindSafe], i32>` at /root/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libstd/rt.rs:51:25
     = note: inside call to `std::rt::lang_start_internal` at /root/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libstd/rt.rs:67:5
     = note: inside call to `std::rt::lang_start::<()>`
```

### PR DESCRIPTION
**Pull request by pnkfelix**
Hide any niche of T from type-construction context of `UnsafeCell<T>`.

Fix #68303 
Fix #68206

### CODE REVIEW

**Comment by Centril** [comment]
cc @RalfJung @rkruppe

File: src/librustc/ty/layout.rs [hunk 1]
```diff
@@ -356,12 +356,14 @@ impl<'tcx> LayoutCx<'tcx, TyCtxt<'tcx>> {
...
             offsets[i as usize] = offset;
 
-            if let Some(mut niche) = field.largest_niche.clone() {
-                let available = niche.available(dl);
-                if available > largest_niche_available {
-                    largest_niche_available = available;
-                    niche.offset += offset;
-                    largest_niche = Some(niche);
+            if !repr.hide_niche() {
+                if let Some(mut niche) = field.largest_niche.clone() {
```
Comment (pnkfelix) [inline src/librustc/ty/layout.rs]: I am aware there are other ways to phrase this that avoid injecting rightward drift, like `let largest_niche = if repr.hide_niche() { None } else { field.largest_niche.clone() }; if let Some(mut niche) = largest_niche { ... }`

if there's demand to do something like that, I'll put it in.

Comment (Centril) [reply src/librustc/ty/layout.rs]: Imo the problem is the language (a lack of let chains) and not your code here. Although adding comments re. what the code does could help?

**Comment by eddyb** [comment]
Was the choice of `#[repr(no_niche)]` discussed, as opposed to making *just* `UnsafeCell` special?
I might be missing something, but I don't think there are other planned uses for `#[repr(no_niche)]`.

**Comment by pnkfelix** [comment]
> Was the choice of `#[repr(no_niche)]` discussed, as opposed to making _just_ `UnsafeCell` special?

We did discuss it at [the relevant lang team meeting](https://github.com/rust-lang/lang-team/blob/master/minutes/2020-01-16.md). I think it is fair to say that no one there was a strong champion of the `#[repr(no_niche)]` approach.

I nonetheless went with it.

It is more readily adaptable if we discover other cases that should be treated like this.

It is also more flexible: for example, if we decide that we want `Cell<T>` to expose its niche, then we could rewrite the stdlib this way, without having to modify `rustc`:

```rust
#[lang = "unsafe_cell"]
#[repr(transparent)]
struct UnsafeCellWithExposedNiche<T: ?Sized> {
    value: T,
}

#[repr(transparent)]
#[repr(no_niche)]
pub struct UnsafeCell<T: ?Sized> {
    value: UnsafeCellWithExposedNiche<T>,
}

#[repr(transparent)]
pub struct Cell<T: ?Sized> {
    value: UnsafeCellWithExposedNiche<T>,
}
```

**Review by oli-obk — APPROVED** [review]

File: src/librustc/ty/layout.rs (hunk 1 above)
Comment (nikomatsakis) [reply src/librustc/ty/layout.rs]: IMO the current structure is clearer than the alternative and the rightward drift is no big deal.

**Comment by nikomatsakis** [comment]
I didn't give a super detailed review, but this looks good to me. I think that having `repr(no_niche)` seems fine, as long as it's considered a pure implementation detail of rustc for now.

File: src/librustc/ty/layout.rs (hunk 1 above)
Comment (eddyb) [reply src/librustc/ty/layout.rs]: Bikeshed alternatives:
```rust
if let (false, Some(mut niche)) = (repr.hide_niche(), field.largest_niche.clone()) {
```
```rust
if let Some(mut niche) = field.largest_niche.clone().filter(|_| !repr.hide_niche()) {
```

Comment (eddyb) [reply src/librustc/ty/layout.rs]: Wait, what am I talking about, the body of this is tiny, nevermind.

File: src/librustc/ty/layout.rs [hunk 2]
```diff
@@ -863,6 +869,11 @@ impl<'tcx> LayoutCx<'tcx, TyCtxt<'tcx>> {
                     return Ok(tcx.intern_layout(st));
                 }
 
+                // At this point, we have handled all unions and
+                // structs. (We have also handled univariant enums
+                // that allow niche-optimization.)
```
Comment (eddyb) [inline src/librustc/ty/layout.rs]: Not sure why niches are mentioned, it's all univariant enums which can be "optimized" (irrespective of niches), i.e. they don't have `#[repr(C)]` nor `#[repr(iN)]` attributes which mandate a certain tagged `enum` layout.

Comment (pnkfelix) [reply src/librustc/ty/layout.rs]: fair enough, I'll `sed -e s/niche-optimization/representation optimization/` here. I just didn't want to accidentally imply that *all* enums are handled in the remaining code.

File: src/librustc/ty/layout.rs (hunk 1 above)
Comment (pnkfelix) [reply src/librustc/ty/layout.rs]: (oh, believe you me, I spent a while trying to convince myself that `if let (false, Some(mut niche)) = (repr.hide_niche(), ...) { ... }` was the way to go. But I just couldn't deal with having the `false` be that "far" from the `hide_niche()` method call. I even considered making an `enum NicheHidden { ... }` to make the bit's meaning more apparent.)

**Comment by pnkfelix** [comment]
(I lost the race with PR #68122; I'll rebase atop current master and update the test to remove `#![feature(transparent_enums)]`)

**Comment by pnkfelix** [comment]
👉 "who's got two thumbs and thought 'surely nothing bad could happen if I just change the issue number in the feature meta-data to point at this PR rather than the issue!' ...? This guy!" 👈

**Comment by nikomatsakis** [comment]
@pnkfelix I think I would prefer to have a true tracking issue, myself, versus pointing to this PR. It'd be great to link to (e.g.) the write-up that I did, the original issue, and to give some context about how this mechanism is rustc-internal and not expected to be stabilized.

**Comment by Centril** [comment]
@nikomatsakis The standard approach to internal feature gates is to put them in the "internal feature gates" section without a tracking issue:
https://github.com/rust-lang/rust/blob/b181835a6bacfa449f55d46764a10e25d1c471dc/src/librustc_feature/active.rs#L96-L98

**Comment by pnkfelix** [comment]
> I think I would prefer to have a true tracking issue, myself, versus pointing to this PR

Okay, I'll file a tracking issue (and do a proper local test run this time instead of assuming I got it right via `x.py test tidy`)

**Comment by pnkfelix** [comment]
> The standard approach to internal feature gates is to put them in the "internal feature gates" section without a tracking issue:

Or oh, maybe I won't then.

**Comment by pnkfelix** [comment]
> @nikomatsakis The standard approach to internal feature gates is to put them in the "internal feature gates" section without a tracking issue:
> 
> https://github.com/rust-lang/rust/blob/b181835a6bacfa449f55d46764a10e25d1c471dc/src/librustc_feature/active.rs#L96-L98

@Centril I did put it in the "// feature-group-start: internal feature gates" section this time around (or at least that is what I think I did); but I also saw examples of some features in that sections that *do* have issue numbers, so I figured I was better off including it ... 

at this point since this is internal only, I guess I will leave it out, since I am worried that making a tracking issue for it will mislead people into thinking it is a feature on the path to stabilization (even if the description for the tracking issue explicitly says otherwise).

**Comment by Centril** [comment]
@pnkfelix Yeah so the tidy mechanism and the grouping used there is newer than most of the internal features, which already happened to have tracking issues, so I had to retcon them into the new system. But newer things don't need to have tracking issues imo. I think your concern re. misleading people that it might be stabilized one day, even with an explicit note to the contrary, is *well founded*. I've seen people ignore such prominent notes in the past with "What's the status of this?" comments before.

**Comment by Dylan-DPC** [comment]
Failed in rollup 

@bors r-

**Comment by jethrogb** [comment]
```
error[E0512]: cannot transmute between types of different sizes, or dependently-sized types
  --> src/libstd/sys/sgx/rwlock.rs:16:5
   |
16 |     mem::transmute::<RWLock, [u8; 128]>(r);
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: source type: `sys::sgx::rwlock::RWLock` (1152 bits)
   = note: target type: `[u8; 128]` (1024 bits)
```

Working on a fix

**Comment by jethrogb** [comment]
Fix in https://github.com/pnkfelix/rust/pull/1

**Comment by pnkfelix** [comment]
Thanks for the patch @jethrogb !

**Comment by pnkfelix** [comment]
looks like the fix doesn't quite work ... 

```
2020-02-03T18:39:05.8494339Z In file included from /tmp/x86_64-fortanix-unknown-sgx_temp/llvm-project/libunwind/src/libunwind.cpp:34:0:
2020-02-03T18:39:05.8495103Z /tmp/x86_64-fortanix-unknown-sgx_temp/llvm-project/libunwind/src/UnwindCursor.hpp: In instantiation of 'RWLock libunwind::DwarfFDECache<libunwind::LocalAddressSpace>::_lock':
2020-02-03T18:39:05.8496130Z /tmp/x86_64-fortanix-unknown-sgx_temp/llvm-project/libunwind/src/UnwindCursor.hpp:177:3: required from 'static void libunwind::DwarfFDECache<A>::iterateCacheEntries(void (*)(unw_word_t, unw_word_t, unw_word_t, unw_word_t)) [with A = libunwind::LocalAddressSpace; unw_word_t = long unsigned int]'
2020-02-03T18:39:05.8496938Z /tmp/x86_64-fortanix-unknown-sgx_temp/llvm-project/libunwind/src/libunwind.cpp:321:37: required from here
2020-02-03T18:39:05.8497406Z /tmp/x86_64-fortanix-unknown-sgx_temp/llvm-project/libunwind/src/UnwindCursor.hpp:96:18: error: too many initializers for 'RWLock'
2020-02-03T18:39:05.8497557Z pthread_rwlock_t DwarfFDECache<A>::_lock = PTHREAD_RWLOCK_INITIALIZER;
2020-02-03T18:39:05.8497815Z ^~~~~~~~~~~~~~~~
2020-02-03T18:39:05.9671696Z make[3]: *** [src/CMakeFiles/unwind_objects.dir/libunwind.cpp.o] Error 1
2020-02-03T18:39:05.9672135Z src/CMakeFiles/unwind_objects.dir/build.make:62: recipe for target 'src/CMakeFiles/unwind_objects.dir/libunwind.cpp.o' failed
2020-02-03T18:39:05.9672465Z CMakeFiles/Makefile2:197: recipe for target 'src/CMakeFiles/unwind_objects.dir/all' failed
2020-02-03T18:39:05.9672761Z make[2]: *** [src/CMakeFiles/unwind_objects.dir/all] Error 2
2020-02-03T18:39:05.9673053Z CMakeFiles/Makefile2:172: recipe for target 'src/CMakeFiles/unwind_static.dir/rule' failed
2020-02-03T18:39:05.9673819Z make[1]: *** [src/CMakeFiles/unwind_static.dir/rule] Error 2
2020-02-03T18:39:05.9674520Z Makefile:190: recipe for target 'unwind_static' failed
2020-02-03T18:39:05.9674810Z make: *** [unwind_static] Error 2
2020-02-03T18:39:13.0779823Z The command '/bin/sh -c /tmp/build-x86_64-fortanix-unknown-sgx-toolchain.sh "63b1b1c5b1ca008099095aa0f8a081a80c8d27f4"' returned a non-zero code: 2
```

is this an artifact of [the below change to the Dockerfile](https://github.com/rust-lang/rust/pull/68491/commits/89975719c563bc874b6dcfb8de250b4b59f7ef4c#diff-560856e4084c86f5bb466571ef99b6f9), which I rubber-stamped?

```diff
  COPY dist-various-2/build-x86_64-fortanix-unknown-sgx-toolchain.sh /tmp/
  # We pass the commit id of the port of LLVM's libunwind to the build script.
  # Any update to the commit id here, should cause the container image to be re-built from this point on.
- RUN /tmp/build-x86_64-fortanix-unknown-sgx-toolchain.sh "53b586346f2c7870e20b170decdc30729d97c42b"
+ RUN /tmp/build-x86_64-fortanix-unknown-sgx-toolchain.sh "63b1b1c5b1ca008099095aa0f8a081a80c8d27f4"

  COPY dist-various-2/build-wasi-toolchain.sh /tmp/
  RUN /tmp/build-wasi-toolchain.sh
```

**Comment by pnkfelix** [comment]
cc @jethrogb regarding above question ^

File: src/ci/docker/dist-various-2/Dockerfile [hunk 3]
```diff
@@ -48,7 +48,7 @@ RUN /tmp/build-solaris-toolchain.sh sparcv9 sparcv9 solaris-sparc
 COPY dist-various-2/build-x86_64-fortanix-unknown-sgx-toolchain.sh /tmp/
 # We pass the commit id of the port of LLVM's libunwind to the build script.
 # Any update to the commit id here, should cause the container image to be re-built from this point on.
-RUN /tmp/build-x86_64-fortanix-unknown-sgx-toolchain.sh "53b586346f2c7870e20b170decdc30729d97c42b"
+RUN /tmp/build-x86_64-fortanix-unknown-sgx-toolchain.sh "63b1b1c5b1ca008099095aa0f8a081a80c8d27f4"
```
Comment (mzohreva) [inline src/ci/docker/dist-various-2/Dockerfile]: 5125c169b30837208a842f85f7ae44a83533bd0e

Comment (pnkfelix) [reply src/ci/docker/dist-various-2/Dockerfile]: This whole thing strikes me as quite fragile ...

**Pull request merged**